src/adera/ShipResources.cpp

```diff
@@ -24,11 +24,40 @@
  */
 #include "ShipResources.h"
 #include "osp/Active/ActiveScene.h"
+#include "osp/Resource/PrototypePart.h"
 
 using namespace osp;
 using namespace osp::active;
 using namespace adera::active::machines;
 
+/* ShipResourceType */
+
+double ShipResourceType::resource_volume(uint64_t quantity) const
+{
+    double units = static_cast<double>(quantity) / std::pow(2.0, m_quanta);
+    return units * m_volume;
+}
+
+double ShipResourceType::resource_mass(uint64_t quantity) const
+{
+    double units = static_cast<double>(quantity) / std::pow(2.0, m_quanta);
+    return units * m_mass;
+}
+
+uint64_t ShipResourceType::resource_capacity(double volume) const
+{
+    double units = volume / m_volume;
+    double quantaPerUnit = std::pow(2.0, m_quanta);
```

Comment:
if you're always turning m_quanta into 2^m_quanta, why even have m_quanta?

Reply:
It's tied to the number of bits used to represent the base unit of resource. You could make an enum or something instead but you can't just pick any old quantaPerUnit.

I will concede though that the language is really confusing in all of this stuff. "Quanta" and "unit" are both super ambiguous words in this context; "quanta" might have to change so that it's easier to differentiate the "unit" 1 and the arbitrary unit chosen by the user.

Reply:
Hmm, now that I think of it, m_quanta is there to increase the precision of the other parameters. If m_quanta is decremented, then the mass, volume, and density can be divided by two to compensate. Though it is pretty useful to describe resources as "1.0m^3 of LOX is 1141.0kg, the game specifically divides this into 65536 units."

m_quanta can just be stored as some unsigned int instead, and is enforced to be a power of two.

A better name might be long but more obvious one: "quantaPerOneUnit"